Stop using _wire setter and use map_wires instead.

AlbertMitjans opened the pull request:
1. Remove all the `_wires` properties.
2. Stop using `op._wires = new_wires`, use `qml.map_wires(op, wire_map=dict(zip(op.wires, new_wires)))` instead.
3. Deprecate `remap_tape_wires`

albi3ro commented on pennylane/transforms/qcut.py:
```diff
@@ -2207,20 +2202,9 @@ def remap_tape_wires(tape: QuantumTape, wires: Sequence) -> QuantumTape:
-            new_wires = Wires([wire_map[w] for w in op.wires])
-            op._wires = new_wires
-            apply(op)
+            qml.map_wires(op, wire_map=wire_map, queue=True)
```
In a follow-up PR (or this one) we could just deprecate/ delete `remap_tape_wires`.  Don't need this function anymore when we can do:
```
qml.map_wires(tape, dict(zip(tape.wires, wires)))
```

AlbertMitjans replied:
Done!

albi3ro commented on pennylane/transforms/optimization/pattern_matching.py:
```diff
@@ -244,28 +243,24 @@ def circuit():
-                            inst._wires = Wires(wires)
-
-                            adjoint(apply, lazy=False)(inst)
+                            inst = qml.map_wires(inst, wire_map=dict(zip(inst.wires, wires)))
+                            inst = qml.map_wires(inst, wire_map=inverse_wires_map)
```
?

albi3ro commented on pennylane/transforms/optimization/pattern_matching.py:
```diff
@@ -228,7 +227,7 @@ def circuit():
                             node = circuit_dag.get_node(elem)
                             inst = copy.deepcopy(node.op)
-                            apply(inst)
+                            qml.map_wires(inst, wire_map=inverse_wires_map, queue=True)
```
Looks like the wires weren't mapped previously. Why are they being mapped now?

albi3ro commented on pennylane/transforms/optimization/pattern_matching.py:
```diff
@@ -244,28 +243,24 @@ def circuit():
                         node = circuit_dag.get_node(node_id)
                         inst = copy.deepcopy(node.op)
-                        apply(inst)
+                        qml.map_wires(inst, wire_map=inverse_wires_map, queue=True)
```
Same here. Why are they being mapped now, when the weren't being mapped before?

albi3ro reviewed (changes requested):
1) Changelog
2) Think there might be some merge artifacts
3) Little confused by the different locations of wire mapping in the pattern matching PR
4) Should test that the warning is emitted by `remap_tape_wires` with `pytest.warns`

AlbertMitjans commented on pennylane/ops/functions/map_wires.py:
```diff
@@ -35,8 +35,6 @@ def map_wires(
 
     Args:
-        input (.Operator, pennylane.QNode, .QuantumScript, or Callable): an operator, quantum node,
-            quantum script, or function that applies quantum operations
```
It was! Fixed

AlbertMitjans commented on pennylane/transforms/optimization/pattern_matching.py:
```diff
@@ -244,28 +243,23 @@ def circuit():
+                tape = qml.map_wires(input=tape_inside, wire_map=inverse_wires_map)
 
     for op in tape.operations:
-        op._wires = Wires([inverse_wires_map[wire] for wire in op.wires.tolist()])
```
Before we ALWAYS needed to use the `inverse_wires_map` because the `commutation_dag` mapped the wires of the `tape` object in-place.

Now, using `qml.map_wires` this doesn't happen, thus we don't need to re-map the wires.

AlbertMitjans commented on pennylane/transforms/optimization/pattern_matching.py:
```diff
@@ -244,28 +243,23 @@ def circuit():
+                        qml.apply(inst)
 
-                tape = tape_inside
+                tape = qml.map_wires(input=tape_inside, wire_map=inverse_wires_map)
```
The re-mapping of the wires using `inverse_wires_map` was done at the end of this method.

However, for the reason explained above, this now only needs to be done when the tape is re-defined.

AlbertMitjans commented:
> 1. Changelog

Done
> 2. Think there might be some merge artifacts

Fixed
> 3. Little confused by the different locations of wire mapping in the pattern matching PR

Some of the changes were wrong, so thank you for catching this! Now I fixed the `commutation_dag.py` file and the changes will make more sense. I also added some comments to clarify the final changes.

Before this refactor, the `CommutationDAG` class modified the tape in-place
> 4. Should test that the warning is emitted by `remap_tape_wires` with `pytest.warns`

Done

albi3ro commented on doc/releases/changelog-dev.md:
```diff
@@ -319,6 +324,9 @@ keyword argument when using `GellMann`, which determines which of the 8 Gell-Man
   composite operator.
   [(#3204)](https://github.com/PennyLaneAI/pennylane/pull/3204)
 
+* Deprecate `remap_tape_wires`. Use `qml.map_wires` instead.
```
```suggestion
* Deprecate `qml.transforms.qcut.remap_tape_wires`. Use `qml.map_wires` instead.
```

albi3ro commented on pennylane/transforms/qcut.py:
```diff
@@ -448,12 +448,7 @@ def graph_to_tape(graph: MultiDiGraph) -> QuantumTape:
         for op in copy_ops:
-            new_wires = Wires([wire_map[w] for w in op.wires])
-
-            # TODO: find a better way to update operation wires
```
Always excellent to remove a `TODO` :)

albi3ro reviewed (approved):
Always love removing `TODO`'s :)
